lib/crypto/src/poseidon2/mod.rs

```diff
@@ -0,0 +1,294 @@
+            index: P::CAPACITY,
+        }
+    }
+
+    /// Size of poseidon sponge's state.
+    #[must_use]
+    pub const fn state_size() -> usize {
+        P::T
+    }
+
+    /// Start index of partial rounds.
+    #[must_use]
+    const fn partial_round_start() -> usize {
+        P::ROUNDS_F / 2
+    }
+
+    /// End index of partial rounds (noninclusive).
+    #[must_use]
+    const fn partial_round_end() -> usize {
+        Self::partial_round_start() + P::ROUNDS_P
+    }
+
+    /// Total number of rounds.
+    #[must_use]
+    const fn rounds() -> usize {
+        P::ROUNDS_F + P::ROUNDS_P
+    }
+
+    /// Absorb a single element into the sponge.
+    ///
+    /// # Panics
+    ///
+    /// May panic if absorbing while squeezing.
+    pub fn absorb(&mut self, elem: &F) {
+        if let Mode::Squeezing = self.mode {
+            panic!("cannot absorb while squeezing");
+        }
+
+        if self.index == Self::state_size() {
+            self.permute();
+            self.index = P::CAPACITY;
+        }
+
+        self.state[self.index] += elem;
+        self.index += 1;
+    }
+
+    /// Absorb batch of elements into the sponge.
+    pub fn absorb_batch(&mut self, elems: &[F]) {
+        for elem in elems {
+            self.absorb(elem);
+        }
+    }
+
+    /// Permute elements in the sponge.
+    pub fn permute(&mut self) {
+        // Linear layer at the beginning.
+        self.matmul_external();
+
+        // Run the first half of the full round.
+        for round in 0..Self::partial_round_start() {
+            self.add_rc_external(round);
+            self.apply_sbox_external();
+            self.matmul_external();
+        }
+
+        // Run the partial round.
+        for round in Self::partial_round_start()..Self::partial_round_end() {
+            self.add_rc_internal(round);
+            self.apply_sbox_internal();
+            self.matmul_internal();
+        }
+
+        // Run the second half of the full round.
+        for round in Self::partial_round_end()..Self::rounds() {
+            self.add_rc_external(round);
+            self.apply_sbox_external();
+            self.matmul_external();
+        }
+    }
+
+    /// Squeeze a single element from the sponge.
+    pub fn squeeze(&mut self) -> F {
+        if matches!(self.mode, Mode::Absorbing)
+            || self.index == Self::state_size()
+        {
+            self.permute();
+            self.mode = Mode::Squeezing;
+            self.index = P::CAPACITY;
+        }
+
+        let elem = self.state[self.index];
+        self.index += 1;
+        elem
+    }
+
+    /// Squeeze a batch of elements from the sponge.
+    pub fn squeeze_batch(&mut self, n: usize) -> Vec<F> {
+        (0..n).map(|_| self.squeeze()).collect()
+    }
+
+    /// Apply sbox to the entire state in the external round.
+    fn apply_sbox_external(&mut self) {
+        for elem in &mut self.state {
+            *elem = elem.pow(P::D);
```

Comment:
Can it fail? Like overflow?

Reply:
rising in `pow` won't fail, since the hole point of finite field (fp) to do operations by modulus.
e.g. `4*4 mod 11 == 16 mod 11 == 3 mod 11`
So you can do many different math operations over `fp` to generate a hash with a good distribution, but still consistent.